[Merged by Bors] - feat(FiberedCategory/HomLift): definition of IsHomLift

callesonne opened the pull request:
We define the notion `IsHomLift` which aims to provide API for working with equalities `p(\phi) = f` for a functor `p` and morphisms `\phi`, `f`. This API is extensively used in an upcoming PR defining the notion of fibered categories. See the repository [Stacks-project](https://github.com/Paul-Lez/Stacks-project) for WIP using this notion.

Co-authored-by: Paul Lezau <paul.lezeau@gmail.com>

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
````diff
@@ -0,0 +1,210 @@
+  R --f--> S
+``` -/
+structure IsHomLift (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) : Prop where
+  (ObjLiftDomain : p.obj a = R)
````
```suggestion
  domain_eq : p.obj a = R
```
The syntax without the braces is prefered. Also, as this field is a "lemma", the initial should not be capitalized. The tactic `by aesop_cat` could be added as default so as to ease automation.

Did you consider making `IsHomLift` a type-class? It seems reasonable to think that it may be convenient. I would suggest the following:

```lean
class Functor.IsHomLift (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) : Prop where
  domain_eq : p.obj a = R := by aesop_cat
  codomain_eq : p.obj b = S := by aesop_cat
  fac : f = eqToHom domain_eq.symm ≫ p.map φ ≫ eqToHom codomain_eq := by aesop_cat

instance (p : 𝒳 ⥤ 𝒮) {a b : 𝒳} (f : a ⟶ b) : p.IsHomLift (p.map f) f where
instance (p : 𝒳 ⥤ 𝒮) (a : 𝒳) : p.IsHomLift (𝟙 (p.obj a)) (𝟙 a) where
```

callesonne replied:
Thanks! I will experiment a bit with making it a class

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,216 @@
+class Functor.IsHomLift (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) : Prop where
+  domain_eq : p.obj a = R := by aesop_cat
+  codomain_eq : p.obj b = S := by aesop_cat
+  commsq : CommSq (p.map φ) (eqToHom domain_eq) (eqToHom codomain_eq) f
```
In order to ease automation, just stating an equality as I suggested is the simpler way. Then, you may introduce lemmas expressing the different variants (`top ≫right = left ≫bottom`, `top = ...`, `bottom = ...`), including a `commSq` lemma and an alternative constructor for `IsHomLift` which would take a `commSq` as a parameter.

callesonne replied:
Okay, that's a great point. Will fix that!

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,216 @@
+  comp_id (𝟙 R) ▸ hφ.comp
+
+/-- If `φ : a ⟶ b` lifts `f` and `ψ : b ⟶ c` lifts `𝟙 T`, then `φ  ≫ ψ` lifts `f` -/
+lemma comp_lift_id {p : 𝒳 ⥤ 𝒮} {R S T : 𝒮} {a b c : 𝒳} {f : R ⟶ S}
+    {φ : a ⟶ b} [hφ : IsHomLift p f φ] {ψ : b ⟶ c} [hψ : IsHomLift p (𝟙 T) ψ] :
+    IsHomLift p f (φ ≫ ψ) where
+  domain_eq := hφ.domain_eq
+  codomain_eq := by rw [hψ.codomain_eq, ← hψ.domain_eq, hφ.codomain_eq]
+  commsq := ⟨by simp only [map_comp, hψ.hom_eq', id_comp, eqToHom_trans, assoc, hφ.3.1]⟩
```
The idea of using a typeclass is that the parameter should no longer need a name `hφ`. You should state lemmas like `domain_eq_of_isHomLift` which would take `p`, `f`, `φ`, and `[IsHomLift p f φ]` as parameters and would produce the expected equality. Then, instead of `hφ.domain_eq`, you would write `domain_eq_of_isHomLift p f φ`.

callesonne replied:
Should I never do `hφ.domain_eq`? I get that it adds automation, so I should definitely add the simp lemma `domain_eq_of_isHomLift`, but I fell that  `hφ.domain_eq` is a lot shorter than `domain_eq_of_isHomLift p f φ`

joelriou replied:
You may try to find a shorter name instead of `domain_eq_of_isHomLift`: it could be a very short name inside a suitable namespace!

callesonne commented:
Okay, I've made some changes now. One thing to note is that the file is significantly slower than before on my computer, but maybe this doesn't matter so much. Another thing that would be nice, but I'm not sure how to add, is if `aesop_cat` (or simp) could deduce `domain_eq` and `codomain_eq` given that a suitable instance of `IsHomLift` is available. That would be helpful for automation when defining  `IsHomLift` instances. For example, I feel that the `domain_eq` field of `comp_lift_id_right` should be able to be automated. Do you know if this is possible in any way?

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,244 @@
+
+open CategoryTheory Category
+
+variable {𝒮 : Type u₁} {𝒳 : Type u₂} [Category 𝒳] [Category 𝒮]
```
```suggestion
variable {𝒮 : Type u₁} {𝒳 : Type u₂} [Category.{v₁} 𝒳] [Category.{v₂} 𝒮]
```

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,244 @@
+
+namespace CategoryTheory
+
+/-- The proposition that an arrow a --φ--> b lifts an arrow R --f--> S in 𝒮 via p. This is
```
The docstring should make more explicit in which category `φ` is and what are the source/target categories for `p`.

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,244 @@
+@[simp]
+protected lemma id {p : 𝒳 ⥤ 𝒮} {R : 𝒮} {a : 𝒳} (ha : p.obj a = R) : p.IsHomLift (𝟙 R) (𝟙 a) where
+
+lemma domain_eq (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b)
+    [p.IsHomLift f φ] : p.obj a = R :=
+  Functor.IsHomLift.domain_eq f φ
```
As the first five lemmas have the same parameters, you could define them in a section:

```suggestion

section

variable (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) [p.IsHomLift f φ]

lemma domain_eq  : p.obj a = R :=
  Functor.IsHomLift.domain_eq f φ
```

joelriou replied:
Also, `p : 𝒳 ⥤ 𝒮` could be defined as a variable more globally in this file.

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,244 @@
+      p.IsHomLift f φ where
+  fac := by simp only [h.1, eqToHom_trans_assoc, eqToHom_refl, id_comp]
+
+lemma eq_of_isHomLift {p : 𝒳 ⥤ 𝒮} (a b : 𝒳) {f : p.obj a ⟶ p.obj b} {φ : a ⟶ b}
+    [p.IsHomLift f φ] : f = p.map φ := by
+  simpa only [eqToHom_refl, comp_id, id_comp] using fac p f φ
```
This seems more appropriate to me:
```suggestion
lemma eq_of_isHomLift (p : 𝒳 ⥤ 𝒮) {a b : 𝒳} (f : p.obj a ⟶ p.obj b) (φ : a ⟶ b)
    [p.IsHomLift f φ] : f = p.map φ := by
  simpa only [eqToHom_refl, comp_id, id_comp] using fac p f φ
```

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,244 @@
+instance self (p : 𝒳 ⥤ 𝒮) {a b : 𝒳} (φ : a ⟶ b) : p.IsHomLift (p.map φ) φ where
+instance (p : 𝒳 ⥤ 𝒮) (a : 𝒳) : p.IsHomLift (𝟙 (p.obj a)) (𝟙 a) where
+
+@[simp]
```
This cannot be a `simp` lemma, because Lean cannot spontaneously find the assumption `p.obj a = R`.
```suggestion
```

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,244 @@
+    [p.IsHomLift (𝟙 R) φ] [p.IsHomLift (𝟙 R) ψ] : p.IsHomLift (𝟙 R) (φ ≫ ψ) :=
+  comp_id (𝟙 R) ▸ comp φ ψ
+
+/-- If `φ : a ⟶ b` lifts `f` and `ψ : b ⟶ c` lifts `𝟙 T`, then `φ  ≫ ψ` lifts `f` -/
+lemma comp_lift_id_right {p : 𝒳 ⥤ 𝒮} {R S T : 𝒮} {a b c : 𝒳} {f : R ⟶ S}
+    {φ : a ⟶ b} [p.IsHomLift f φ] {ψ : b ⟶ c} [p.IsHomLift (𝟙 T) ψ] :
+    p.IsHomLift f (φ ≫ ψ) where
+  -- TODO: this first one should be able to be automated?
+  domain_eq := domain_eq p f φ
+  codomain_eq := by rw [codomain_eq p (𝟙 T) ψ, ← domain_eq p (𝟙 T) ψ, codomain_eq p f φ]
+  fac := by simp [fac p f φ, fac' p (𝟙 T) ψ]
```
```suggestion
/-- If `φ : a ⟶ b` lifts `f` and `ψ : b ⟶ c` lifts `𝟙 T`, then `φ  ≫ ψ` lifts `f` -/
lemma comp_lift_id_right (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b c : 𝒳} (f : R ⟶ S)
    (φ : a ⟶ b) [p.IsHomLift f φ] (ψ : b ⟶ c) (T : 𝒮) [p.IsHomLift (𝟙 T) ψ] :
    p.IsHomLift f (φ ≫ ψ) := by
  obtain rfl : S = T := by rw [← codomain_eq p f φ, domain_eq p (𝟙 T) ψ]
  simpa using inferInstanceAs (p.IsHomLift (f ≫ 𝟙 S) (φ ≫ ψ))
 ```

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,244 @@
+  fac := by simp [fac p f φ, fac' p (𝟙 T) ψ]
+
+/-- If `φ : a ⟶ b` lifts `𝟙 T` and `ψ : b ⟶ c` lifts `f`, then `φ  ≫ ψ` lifts `f` -/
+lemma comp_lift_id_left {p : 𝒳 ⥤ 𝒮} {R S T : 𝒮} {a b c : 𝒳} {f : S ⟶ T}
```
Same as above.

joelriou commented:
> One thought I just had, is that maybe it could be a good idea to add the class `IsObjLift`, which would allow type-class inference to infer `domain_eq` and `codomain_eq`. But maybe this is a bit overkill.

I have added a suggestion of alternate proof for `comp_lift_id_right`. If we can be avoid introducing `IsObjLift`. It is probably better!

(In proofs involving a map `f` over another map `g`, I think that in many cases, we should be able to do substitutions so that we may assume that `g` is definitionally `p.map f`: first substitute the equality of objects (as I did above), then `simp` the map equality to get rid of the `eqToHom`, and then substitute the equality of maps.)

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,233 @@
+  rw [p.map_comp]
+  apply CommSq.horiz_comp (commSq p f φ) (commSq p g ψ)
+
+/-- If `φ : a ⟶ b` and `ψ : b ⟶ c` lift `𝟙 S`, then so does `φ ≫ ψ` -/
```
```suggestion
/-- If `φ : a ⟶ b` and `ψ : b ⟶ c` lift `𝟙 R`, then so does `φ ≫ ψ` -/
```

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,233 @@
+  comp_id f ▸ comp φ (eqToHom h)
+
+instance lift_eqToHom_comp {R' R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) (h : R' = R)
+    [p.IsHomLift f φ] : p.IsHomLift ((eqToHom h) ≫ f) φ :=
```
```suggestion
    [p.IsHomLift f φ] : p.IsHomLift (eqToHom h ≫ f) φ :=
```

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,233 @@
+  id_comp φ ▸ comp (𝟙 a) φ
+
+instance lift_comp_eqToHom {R S S': 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) (h : S = S')
+    [p.IsHomLift f φ] : p.IsHomLift (f ≫ (eqToHom h)) φ :=
```
```suggestion
    [p.IsHomLift f φ] : p.IsHomLift (f ≫ eqToHom h) φ :=
```

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,233 @@
+
+@[simp]
+lemma lift_eqToHom_comp_iff {R' R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) (h : R' = R) :
+    p.IsHomLift ((eqToHom h) ≫ f) φ ↔ p.IsHomLift f φ where
```
```suggestion
    p.IsHomLift (eqToHom h ≫ f) φ ↔ p.IsHomLift f φ where
```

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,233 @@
+
+@[simp]
+lemma lift_comp_eqToHom_iff {R S S' : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) (h : S = S') :
+    p.IsHomLift (f ≫ (eqToHom h)) φ ↔ p.IsHomLift f φ where
```
```suggestion
    p.IsHomLift (f ≫ eqToHom h) φ ↔ p.IsHomLift f φ where
```

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,233 @@
+  mpr := fun hφ => inferInstance
+
+/-- The isomorphism `R ≅ S` obtained from an isomorphism `φ : a ≅ b` lifting `f` -/
+def isoOfIsoLift  {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ≅ b) [p.IsHomLift f φ.hom] :
```
I would suggest you define direclty both the `hom` and `inv` fields of this `isoOfIsoLift`, so that the proof of `isoOfIsoLift_hom` becomes `rfl`. This may facilitate some future proofs, because `dsimp` would already be able to simplify.

callesonne commented:
I have now fixed your comments, as well as redefined `IsHomLift` using inductive types as in the zulip thread. I also added a macro `subst_hom_lift` to make the substitution a bit more readable.

joelriou commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,248 @@
+  mpr := fun hφ => inferInstance
+
+/-- The isomorphism `R ≅ S` obtained from an isomorphism `φ : a ≅ b` lifting `f` -/
+def isoOfIsoLift  {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ≅ b) [p.IsHomLift f φ.hom] :
```
The following syntax would automatically generate the next lemma:
```suggestion
@[simps hom]
def isoOfIsoLift  {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ≅ b) [p.IsHomLift f φ.hom] :
```

callesonne replied:
Perfect! Thanks :)

callesonne commented:
I'm not convinced about using `IsHomLift` as a class anymore, basically for the following two reasons:
1. `IsHomLift p f φ` makes it so that I can leave parameters `p, f, φ` implicit, which makes it slightly more convenient to apply lemmas which has it as a parameter. Especially in situations where the expressions for `f` and `φ` are somewhat complicated (involving `eqToHoms` ) it would be quite convenient to not have to write them down explicitly.
2. There are often situations where type-class inference can't infer `IsHomLift`, and having it as a class makes it quite inconvenient to try and give this instance explicitly. For example the following lemma is very simple, but can't be inferred by type-class inference:
`protected lemma id {p : 𝒳 ⥤ 𝒮} {R : 𝒮} {a : 𝒳} (ha : p.obj a = R) : p.IsHomLift (𝟙 R) (𝟙 a) := `.
So whenever I want to use this instance, I have to either have another line before with a have statement, which feels inconvenient for such a simple lemma, or use `@` notation.

Also, I havn't felt many of the benefits of having it as a class, I think the only noticeable thing to me is that `IsHomLift.comp` can be inferred can be quite nice.

It is very possible that I am not fully using the benefits of having a class (and that these issues shouldn't arise if I use it correctly), so please take what I say with a grain of salt :)

joelriou commented:
>     1. `IsHomLift p f φ` makes it so that I can leave parameters `p, f, φ` implicit, which makes it slightly more convenient to apply lemmas which has it as a parameter. Especially in situations where the expressions for `f` and `φ` are somewhat complicated (involving `eqToHoms` ) it would be quite convenient to not have to write them down explicitly.

The change of design (using an inductive type, the API for substitutions, etc) makes it so that in principle, you should not need to write that many lemmas where `eqToHom` should appear in the assumptions (although, they may occasionally appear in the statements).

>     2. There are often situations where type-class inference can't infer `IsHomLift`, and having it as a class makes it quite inconvenient to try and give this instance explicitly. For example the following lemma is very simple, but can't be inferred by type-class inference:
>        `protected lemma id {p : 𝒳 ⥤ 𝒮} {R : 𝒮} {a : 𝒳} (ha : p.obj a = R) : p.IsHomLift (𝟙 R) (𝟙 a) := `.

We should basically almost never need to use this lemma, because we may do `subst ha`, and then the instance can be infered!

callesonne commented:
Okay, this makes sense. I think the times I have encountered it has all been in short term-mode proofs, when I had to change the proof from 1 to 2 lines because of this. But then like you say this will be the worst case, as in longer proofs I should just do the substitution first. So I am happy with it :)

kbuzzard commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,246 @@
+lemma domain_eq : p.obj a = R := by
+  subst_hom_lift p f φ; rfl
+
+lemma codomain_eq  : p.obj b = S := by
```
```suggestion
lemma codomain_eq : p.obj b = S := by
```

kbuzzard commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,246 @@
+/-- Given a morphism `f : R ⟶ S`, and an isomorphism `φ : a ≅ b` lifting `f`, `isoOfIsoLift f φ` is
+the isomorphism `Φ : R ≅ S` with `Φ.hom = f` induced from `φ` -/
+@[simps hom]
+def isoOfIsoLift  {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ≅ b) [p.IsHomLift f φ.hom] :
```
```suggestion
def isoOfIsoLift {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ≅ b) [p.IsHomLift f φ.hom] :
```

kbuzzard replied:
```suggestion
def isoOfIsoLift {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ≅ b) [p.IsHomLift f φ.hom] :
```

kbuzzard commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,246 @@
+
+lemma of_fac {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) (ha : p.obj a = R) (hb : p.obj b = S)
+    (h : f = eqToHom ha.symm ≫ p.map φ ≫ eqToHom hb) : p.IsHomLift f φ := by
+  subst ha hb
+  obtain rfl : f = p.map φ := by simpa using h
+  infer_instance
```
```suggestion
  subst ha hb h
  simp
```

kbuzzard commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,246 @@
+  obtain rfl : R = S := by rw [← codomain_eq p (𝟙 R) φ, domain_eq p f ψ]
+  simpa using inferInstanceAs (p.IsHomLift (𝟙 R ≫ f) (φ ≫ ψ))
+
+lemma eqToHom_domain_lift_id {p : 𝒳 ⥤ 𝒮} {a b : 𝒳} (hab : a = b) {R : 𝒮} (hR : p.obj a = R) :
```
This and the next three lemmas can be instances. Is there a reason you want to keep them as lemmas?

callesonne replied:
Can they? How will the hypothesis `hR` be inferred?

kbuzzard replied:
Aah of course you're right. Thanks!

kbuzzard commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,246 @@
+
+instance comp_eqToHom_lift {R S : 𝒮} {a' a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) (h : a' = a)
+    [p.IsHomLift f φ] : p.IsHomLift f (eqToHom h ≫ φ) := by
+  subst h; aesop
```
```suggestion
  subst h; simp_all
```
`aesop?` reveals that it's just doing `simp_all` so you may as well do it directly. Similar comment for the next three lemmas.

kbuzzard commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,246 @@
+/-- If `φ : a ⟶ b` lifts `f : R ⟶ S` and both are isomorphisms, then `φ⁻¹` lifts `f⁻¹`. -/
+protected instance inv {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) [IsIso f] [IsIso φ]
+    [p.IsHomLift f φ] : p.IsHomLift (inv f) (inv φ) :=
+  @IsHomLift.inv_lift_inv _ _ _ _ p _ _ _ _ (asIso f) (asIso φ) (by aesop)
```
```suggestion
  have : p.IsHomLift (asIso f).hom (asIso φ).hom := by simp_all
  IsHomLift.inv_lift_inv p (asIso f) (asIso φ)
```
My instinct is to avoid all the @ nonsense.

kbuzzard commented on Mathlib/CategoryTheory/FiberedCategory/HomLift.lean:
```diff
@@ -0,0 +1,246 @@
+lifts `𝟙 S` -/
+instance lift_id_inv (S : 𝒮) {a b : 𝒳} (φ : a ⟶ b) [IsIso φ] [p.IsHomLift (𝟙 S) φ] :
+    p.IsHomLift (𝟙 S) (inv φ) :=
+  (IsIso.inv_id (X:=S)) ▸ (IsHomLift.inv p _ φ)
```
```suggestion
  (IsIso.inv_id (X := S)) ▸ (IsHomLift.inv p _ φ)
```

kbuzzard reviewed (approved):
I've left some superficial comments but this looks good to go to me.

joelriou commented:
Thanks!

bors merge
